From 37093792973ab82413a810cc4201a22a455bd68b Mon Sep 17 00:00:00 2001
From: Linu Cherian <lcherian@marvell.com>
Date: Thu, 23 Jan 2020 16:53:03 +0530
Subject: [PATCH 0669/1921] coresight: etr: Enable hardware quirks based on
 Silicon revision

- Hardware quirks are now enabled based on Silicon revision
- Renamed and moved related macros to a file which is common
  to both ETM and ETR drivers, since there are cases where
  changes are applicable to both ETM and ETR drivers.

Change-Id: I2adec4234edde2f245f4ff69154bc2022fd226c7
Signed-off-by: Linu Cherian <lcherian@marvell.com>
Reviewed-on: https://sj1git1.cavium.com/22245
Reviewed-on: https://sj1git1.cavium.com/c/IP/SW/kernel/linux/+/32244
[WK: The original patch got from Marvell sdk11.21.09]
Signed-off-by: Wenlin Kang <wenlin.kang@windriver.com>
---
 drivers/hwtracing/coresight/Makefile          |  3 ++-
 drivers/hwtracing/coresight/coresight-priv.h  | 10 +++++++
 .../hwtracing/coresight/coresight-quirks.c    | 26 +++++++++++++++++++
 .../hwtracing/coresight/coresight-tmc-etr.c   | 12 ++++-----
 drivers/hwtracing/coresight/coresight-tmc.c   |  9 +++----
 drivers/hwtracing/coresight/coresight-tmc.h   |  8 ------
 6 files changed, 47 insertions(+), 21 deletions(-)
 create mode 100644 drivers/hwtracing/coresight/coresight-quirks.c

diff --git a/drivers/hwtracing/coresight/Makefile b/drivers/hwtracing/coresight/Makefile
index 3c0ac421e211..74e38967f2f6 100644
--- a/drivers/hwtracing/coresight/Makefile
+++ b/drivers/hwtracing/coresight/Makefile
@@ -2,7 +2,8 @@
 #
 # Makefile for CoreSight drivers.
 #
-obj-$(CONFIG_CORESIGHT) += coresight.o coresight-etm-perf.o coresight-platform.o
+obj-$(CONFIG_CORESIGHT) += coresight.o coresight-etm-perf.o \
+			   coresight-platform.o coresight-quirks.o
 obj-$(CONFIG_CORESIGHT_LINK_AND_SINK_TMC) += coresight-tmc.o \
 					     coresight-tmc-etf.o \
 					     coresight-tmc-etr.o
diff --git a/drivers/hwtracing/coresight/coresight-priv.h b/drivers/hwtracing/coresight/coresight-priv.h
index 1294805f2bcb..afc0d29c7969 100644
--- a/drivers/hwtracing/coresight/coresight-priv.h
+++ b/drivers/hwtracing/coresight/coresight-priv.h
@@ -68,6 +68,14 @@ static DEVICE_ATTR_RO(name)
 extern const u32 barrier_pkt[4];
 #define CORESIGHT_BARRIER_PKT_SIZE (sizeof(barrier_pkt))
 
+/* Marvell OcteonTx CN9xxx device */
+#define OCTEONTX_CN9XXX_ETR		0x000cc213
+
+/* Coresight Hardware quirks */
+#define CSETR_QUIRK_BUFFSIZE_8BX	(0x1U << 0) /* 8 byte size multiplier */
+#define CSETR_QUIRK_SECURE_BUFF		(0x1U << 1) /* Trace buffer is Secure */
+#define CSETR_QUIRK_RESET_CTL_REG	(0x1U << 2) /* Reset CTL on reset */
+
 enum etm_addr_type {
 	ETM_ADDR_TYPE_NONE,
 	ETM_ADDR_TYPE_SINGLE,
@@ -204,5 +212,7 @@ static inline void *coresight_get_uci_data(const struct amba_id *id)
 }
 
 void coresight_release_platform_data(struct coresight_platform_data *pdata);
+/* Coresight ETM/ETR hardware quirks */
+u32 coresight_get_etr_quirks(unsigned int id);
 
 #endif
diff --git a/drivers/hwtracing/coresight/coresight-quirks.c b/drivers/hwtracing/coresight/coresight-quirks.c
new file mode 100644
index 000000000000..f7389a29b72d
--- /dev/null
+++ b/drivers/hwtracing/coresight/coresight-quirks.c
@@ -0,0 +1,26 @@
+// SPDX-License-Identifier: GPL-2.0
+//
+#include <asm/cputype.h>
+#include <linux/coresight.h>
+#include "coresight-priv.h"
+
+u32 coresight_get_etr_quirks(unsigned int id)
+{
+	u32 options = 0; /* reset */
+
+	if (midr_is_cpu_model_range(read_cpuid_id(),
+				     MIDR_MRVL_OCTEONTX2_96XX,
+				     MIDR_CPU_VAR_REV(0, 0),
+				     MIDR_CPU_VAR_REV(3, 0)) ||
+	    midr_is_cpu_model_range(read_cpuid_id(),
+				     MIDR_MRVL_OCTEONTX2_95XX,
+				     MIDR_CPU_VAR_REV(0, 0),
+				     MIDR_CPU_VAR_REV(2, 0)))
+		options |= CSETR_QUIRK_BUFFSIZE_8BX |
+			CSETR_QUIRK_RESET_CTL_REG;
+
+	if (id == OCTEONTX_CN9XXX_ETR)
+		options |= CSETR_QUIRK_SECURE_BUFF;
+
+	return options;
+}
diff --git a/drivers/hwtracing/coresight/coresight-tmc-etr.c b/drivers/hwtracing/coresight/coresight-tmc-etr.c
index f5265c687e52..08d46f5c7624 100644
--- a/drivers/hwtracing/coresight/coresight-tmc-etr.c
+++ b/drivers/hwtracing/coresight/coresight-tmc-etr.c
@@ -625,7 +625,7 @@ static int tmc_etr_alloc_flat_buf(struct tmc_drvdata *drvdata,
 		return -ENOMEM;
 	}
 
-	if (!(drvdata->etr_options & CORESIGHT_OPTS_SECURE_BUFF))
+	if (!(drvdata->etr_options & CSETR_QUIRK_SECURE_BUFF))
 		goto skip_secure_buffer;
 
 	/* Register driver allocated dma buffer for necessary
@@ -899,7 +899,7 @@ static struct etr_buf *tmc_alloc_etr_buf(struct tmc_drvdata *drvdata,
 	/* TODO: Consider using etr_buf->secure everywhere instead of
 	 * etr_options for consistency
 	 */
-	if (drvdata->etr_options & CORESIGHT_OPTS_SECURE_BUFF)
+	if (drvdata->etr_options & CSETR_QUIRK_SECURE_BUFF)
 		etr_buf->secure = true;
 
 	/*
@@ -1013,13 +1013,13 @@ static void __tmc_etr_enable_hw(struct tmc_drvdata *drvdata)
 
 	CS_UNLOCK(drvdata->base);
 
-	if (drvdata->etr_options & CORESIGHT_OPTS_RESET_CTL_REG)
+	if (drvdata->etr_options & CSETR_QUIRK_RESET_CTL_REG)
 		tmc_disable_hw(drvdata);
 
 	/* Wait for TMCSReady bit to be set */
 	tmc_wait_for_tmcready(drvdata);
 
-	if (drvdata && CORESIGHT_OPTS_BUFFSIZE_8BX)
+	if (drvdata->etr_options && CSETR_QUIRK_BUFFSIZE_8BX)
 		writel_relaxed(etr_buf->size / 8, drvdata->base + TMC_RSZ);
 	else
 		writel_relaxed(etr_buf->size / 4, drvdata->base + TMC_RSZ);
@@ -1039,7 +1039,7 @@ static void __tmc_etr_enable_hw(struct tmc_drvdata *drvdata)
 		axictl |= TMC_AXICTL_SCT_GAT_MODE;
 
 	writel_relaxed(axictl, drvdata->base + TMC_AXICTL);
-	if (drvdata->etr_options & CORESIGHT_OPTS_SECURE_BUFF)
+	if (drvdata->etr_options & CSETR_QUIRK_SECURE_BUFF)
 		tmc_write_dba(drvdata, etr_buf->s_hwaddr);
 	else
 		tmc_write_dba(drvdata, etr_buf->hwaddr);
@@ -1051,7 +1051,7 @@ static void __tmc_etr_enable_hw(struct tmc_drvdata *drvdata)
 	 */
 	if (tmc_etr_has_cap(drvdata, TMC_ETR_SAVE_RESTORE)) {
 		tmc_write_rrp(drvdata, etr_buf->hwaddr);
-		if (drvdata->etr_options & CORESIGHT_OPTS_SECURE_BUFF)
+		if (drvdata->etr_options & CSETR_QUIRK_SECURE_BUFF)
 			tmc_write_rwp(drvdata, etr_buf->s_hwaddr);
 		else
 			tmc_write_rwp(drvdata, etr_buf->hwaddr);
diff --git a/drivers/hwtracing/coresight/coresight-tmc.c b/drivers/hwtracing/coresight/coresight-tmc.c
index 2fdad6ec21eb..86f18942e520 100644
--- a/drivers/hwtracing/coresight/coresight-tmc.c
+++ b/drivers/hwtracing/coresight/coresight-tmc.c
@@ -183,7 +183,7 @@ static ssize_t tmc_read(struct file *file, char __user *data, size_t len,
 	if (actual <= 0)
 		return 0;
 
-	if ((drvdata->etr_options & CORESIGHT_OPTS_SECURE_BUFF) &&
+	if ((drvdata->etr_options & CSETR_QUIRK_SECURE_BUFF) &&
 		tmc_copy_secure_buffer(drvdata, bufp, actual))
 		return -EFAULT;
 
@@ -467,11 +467,8 @@ static int tmc_probe(struct amba_device *adev, const struct amba_id *id)
 
 	drvdata->cpu = coresight_get_cpu(dev);
 
-	/* Enable options for Silicon issues */
-	if (id->id == OCTEONTX_CN9XXX_ETR)
-		drvdata->etr_options = CORESIGHT_OPTS_BUFFSIZE_8BX |
-					CORESIGHT_OPTS_SECURE_BUFF |
-					CORESIGHT_OPTS_RESET_CTL_REG;
+	/* Enable fixes for Silicon issues */
+	drvdata->etr_options = coresight_get_etr_quirks(OCTEONTX_CN9XXX_ETR);
 
 	devid = readl_relaxed(drvdata->base + CORESIGHT_DEVID);
 	drvdata->config_type = BMVAL(devid, 6, 7);
diff --git a/drivers/hwtracing/coresight/coresight-tmc.h b/drivers/hwtracing/coresight/coresight-tmc.h
index a813581116e6..716ce8c2dcb8 100644
--- a/drivers/hwtracing/coresight/coresight-tmc.h
+++ b/drivers/hwtracing/coresight/coresight-tmc.h
@@ -138,14 +138,6 @@ enum tmc_mem_intf_width {
 #define OCTEONTX_CN9XXX_ETR_CAPS	\
 	(TMC_ETR_SAVE_RESTORE)
 
-/* Marvell OcteonTx CN9xxx device */
-#define OCTEONTX_CN9XXX_ETR		0x000cc213
-
-/* Marvell OcteonTx CN9xxx HW issues */
-#define CORESIGHT_OPTS_BUFFSIZE_8BX	(0x1U << 0) /* 8 byte size multiplier */
-#define CORESIGHT_OPTS_SECURE_BUFF	(0x1U << 1) /* Trace buffer is Secure */
-#define CORESIGHT_OPTS_RESET_CTL_REG	(0x1U << 2) /* Reset CTL on reset */
-
 /* SMC call ids for managing the secure trace buffer */
 
 /* Args: x1 - size, x2 - cpu, x3 - llc lock flag
-- 
2.31.1

